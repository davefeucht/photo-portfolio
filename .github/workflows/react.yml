# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
