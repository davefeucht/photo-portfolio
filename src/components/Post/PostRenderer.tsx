import PostFooter from 'components/PostFooter/PostFooter';
import PostImage from 'components/PostImage/PostImage';
import PostTitlebar from 'components/PostTitlebar/PostTitlebar';
import { reaction } from 'mobx';
import * as React from 'react';
import { useEffect, useRef, useState } from 'react';
import { setPostRect } from 'utils/PostHelper';
import { ScreenInfo, VisiblePost } from 'utils/types';

interface Props {
    onClose(): void;
    previousPost: number;
    nextPost: number;
    screenInfo: ScreenInfo;
    visiblePost: VisiblePost;
}

const image = document.createElement('img');

const PostRenderer: React.FC<Props> = ({
    onClose,
    previousPost,
    nextPost,
    screenInfo,
    visiblePost
}) => {
    const [imageHeight, setImageHeight] = useState<number>(0);
    const imageRef = useRef<HTMLImageElement>(null);

    const updateImage = () => {
        setPostRect(image, screenInfo.width, screenInfo.height, visiblePost);
    };

    image.onload = () => {
        updateImage();
        setImageHeight(imageRef.current?.clientHeight ?? 0);
    };

    useEffect(() => {
        if (visiblePost.fullImageUrl) {
            image.src = visiblePost.fullImageUrl;
        }

        if (imageRef.current) {
            imageRef.current.addEventListener('transitionend', updateImage, true);
        }

        const disposer = reaction(
            () => [screenInfo.width, screenInfo.height],
            () => updateImage()
        );

        updateImage();

        return () => {
            if (imageRef.current) {
                imageRef.current.removeEventListener('transitionend', updateImage);
            }
            disposer();
        };
    }, [visiblePost.fullImageUrl]);

    return (
        <div className="post-background" onClick={onClose}>
            <div className="post" onClick={event => event.stopPropagation()}>
                <PostTitlebar postTitle={visiblePost.postTitle} />
                <PostImage
                    imageUrl={visiblePost.fullImageUrl}
                    imageHeight={imageHeight}
                    previousPost={previousPost}
                    nextPost={nextPost}
                    ref={imageRef}
                />
                <PostFooter tagNames={visiblePost.tagNames} />
            </div>
        </div>
    );
};

export default PostRenderer;
