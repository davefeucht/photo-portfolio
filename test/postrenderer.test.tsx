import { render, screen, waitFor } from '@testing-library/react';
import PostRenderer from 'components/Post/PostRenderer';
import * as React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { visiblePost } from './data/testData';

describe("PostRenderer", () => {
    it("Renders", async () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/post/1']}>
                <PostRenderer
                    onClose={jest.fn()}
                    onImageLoad={jest.fn()}
                    imageHeight={300}
                    previousPost={0}
                    nextPost={2}
                    screenInfo={{ width: 1024, height: 768 }}
                    visiblePost={visiblePost}
                />
            </MemoryRouter>
        );

        await waitFor(() => expect(screen.getByText(visiblePost.postTitle)).toBeInTheDocument());
        expect(container).toMatchSnapshot();
    });

    it('Calls onload when image loads', async () => {
        const onLoad = jest.fn();

        render(
            <MemoryRouter initialEntries={['/post/1']}>
                <PostRenderer
                    onClose={jest.fn()}
                    onImageLoad={onLoad}
                    imageHeight={300}
                    previousPost={0}
                    nextPost={2}
                    screenInfo={{ width: 1024, height: 768 }}
                    visiblePost={visiblePost}
                />
            </MemoryRouter>
        );

        await waitFor(() => expect(screen.getByText(visiblePost.postTitle)).toBeInTheDocument());
        expect(onLoad).toHaveBeenCalledTimes(2);
    });

    it('Calls onload when screen size changes', async () => {
        const onLoad = jest.fn();

        const { rerender } = render(
            <MemoryRouter initialEntries={['/post/1']}>
                <PostRenderer
                    onClose={jest.fn()}
                    onImageLoad={onLoad}
                    imageHeight={300}
                    previousPost={0}
                    nextPost={2}
                    screenInfo={{ width: 1024, height: 768 }}
                    visiblePost={visiblePost}
                />
            </MemoryRouter>
        );

        expect(onLoad).toHaveBeenCalledTimes(2);

        rerender(
            <MemoryRouter initialEntries={['/post/1']}>
                <PostRenderer
                    onClose={jest.fn()}
                    onImageLoad={onLoad}
                    imageHeight={300}
                    previousPost={0}
                    nextPost={2}
                    screenInfo={{ width: 800, height: 600 }}
                    visiblePost={visiblePost}
                />
            </MemoryRouter>
        )

        await waitFor(() => expect(screen.getByText(visiblePost.postTitle)).toBeInTheDocument());
        await waitFor(() => expect(onLoad).toHaveBeenCalledTimes(3));
    });
});