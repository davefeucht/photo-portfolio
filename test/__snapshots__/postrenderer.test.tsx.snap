// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostRenderer Renders 1`] = `
<div>
  <div
    class="post-background"
  >
    <div
      class="post"
    >
      <div
        class="post-titlebar"
      >
        <div
          class="title"
        >
          Post 1
        </div>
        <div
          class="close-button"
          role="button"
        >
          <a
            href="/category/undefined"
          >
            <div
              class="close-button-content"
            >
              <img
                height="30px"
                src="./assets/images/close-button.svg"
                width="30px"
              />
            </div>
          </a>
        </div>
      </div>
      <div
        aria-label="post-image"
        class="post-image"
      >
        <img
          src="https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg"
        />
      </div>
      <div
        class="post-footer"
      >
        <div
          class="labels"
        >
          something, somethingelse
        </div>
      </div>
    </div>
  </div>
</div>
`;
